### PR TITLE
[Support] Add class for generating unique names based on a symbol cache

### PR DESCRIPTION
**Pull request by mortbopet**
This commit introduces a new utility class for generating unique names based on some input base name. The logic for permutating the base name can be user-specified, but defaults to a simple inter-suffixed scheme.

As a sample use, `HWSpecialize` has been rewritten to use the `SymbolUniquer`.

Based on #3101.

### CODE REVIEW

**Comment by mortbopet** [comment]
Having some issues ATM reproducing the CI error locally.

File: include/circt/Support/SymUniquer.h [hunk 1]
```diff
@@ -0,0 +1,76 @@
...
+
+namespace circt {
+
+/// The SymbolUniquer is used to generate unique names based on some set of
+/// existing names defined through a SymbolCache.
+class SymbolUniquer {
+  /// Uniquer function which suffixes a base name with an integer, starting at
+  /// 0.
+  static std::pair<std::string, llvm::Any>
+  integerUniquer(mlir::StringRef name, llvm::Any previousToken) {
```
Comment (teqdruid) [inline include/circt/Support/SymUniquer.h:26]: I'll give this a more thorough review once the symbol cache stuff is moved and is divorced from this PR. Early comment: why `llvm::Any` vs a template? I am wary of any sort of dynamic type behavior.

Comment (mortbopet) [reply include/circt/Support/SymUniquer.h:26]: This design was motivated by not having `SymbolUniquer` be a template, but I do see your point of view. A templated `SymbolUniquer` with a default implementation being "integer-uniqueing" (as shown) could be an alternative.

**Comment by mikeurbach** [comment]
If you make the base branch mortbopet:support/move_symcache, it would make it simpler to discuss the changes in this PR in isolation while the other is outstanding, and GitHub will automatically update the base once the other one merges.

**Comment by teqdruid** [comment]
> If you make the base branch mortbopet:support/move_symcache, it would make it simpler to discuss the changes in this PR in isolation while the other is outstanding, and GitHub will automatically update the base once the other one merges.

IME, that doesn't work properly when the src branches are in a fork.

**Comment by mortbopet** [comment]
> IME, that doesn't work properly when the src branches are in a fork.

Correct - i usually work off my own fork to avoid polluting branches in upstream, but in this case it would have been helpful for the PRs...!

**Comment by mortbopet** [comment]
Closing in favor of #3137.

**Pull request closed without merging**